rfcs/0003-mutations.md

```diff
@@ -0,0 +1,280 @@
+* _Optimistic Response_: A response given to the dApp that predicts what the outcome of the mutation's execution will be. If it is incorrect, it will be overwritten with the actual result.  
+
+## Detailed Design
+
+### Mutation Manifest
+
+`subgraph.yaml`
+```yaml
+specVersion: ...
+...
+mutations:
+  specVersion: 0.0.1
+  repository: https://npmjs.com/package/...
+  schema:
+    file: ./mutations/schema.graphql
+  resolvers:
+    kind: javascript
+    file: ./mutations/index.js
+dataSources: ...
+...
+```
+
+Alternatively, you can store the mutations manifest externally like so:  
+`subgraph.yaml`
+```yaml
+specVersion: ...
+...
+mutations:
+  file: ./mutations/mutations.yaml
+dataSources: ...
+...
+```
+`mutations/mutations.yaml`
+```yaml
+specVersion: 0.0.1
+repository: https://npmjs.com/package/...
+schema:
+  file: ./schema.graphql
+resolvers:
+  kind: javascript
+  file: ./index.js
+```
+
+### Mutation Schema
+`schema.graphql`
+```graphql
+type MyEntity @entity {
+  id: ID!
+  name: String!
+  value: BigInt!
+}
+```
+
+`mutations/schema.graphql`
+```graphql
+input MyEntityOptions {
+  name: String!
+  value: BigInt!
+}
+
+type Mutation {
+  createEntity(
+    options: MyEntityOptions!
+  ): MyEntity!
+
+  setEnityName(
+    entity: MyEntity!
+    name: String!
+  ): MyEntity!
+}
+```
+**NOTE:** GraphQL types from the subgraph's schema.graphql are automatically included in this file.
+
+### Mutation Resolvers
+`mutations/index.js`
+```javascript
+const resolvers = {
+  Mutation: {
+    async createEntity (_, args, context) {
+      // Extract mutation arguments
+      const { name, value } = args.options
+
+      // Use configuration properties created by the
+      // config generator functions below
+      const { ethereum, ipfs } = context.graph.config
+
+      // Fetch datasource addresses & abis
+      const { MyContract } = context.graph.datasources
+      await MyContract.abi
+      await MyContract.address
+
+      // Modify a state object, which relays updates back
+      // to the subscribed dApp
+      const { mutationState } = context.graph
+      mutationState.addTransaction("tx_hash")
+
+      ...
+    },
+    async setEntityName (_, args, context) {
+      ...
+    }
+  }
+}
+
+// Configuration Setters
+const config = {
+  // These function arguments are passed in by the dApp
+  ethereum (provider) {
+    return new ethers.providers.Web3Provider(provider)
+  },
+  ipfs (provider) {
+    return new IPFS(provider)
+  },
+  customProperty (value) {
+    return value + 2
+  }
+}
+
+export default {
+  resolvers,
+  config
+}
+```
+
+### dApp Integration
+```javascript
+const {
+  createMutations,
+  createMutationsLink,
+  useMutation
+} = require("@graphprotocol/mutations-ts")
+const myMutations = require("mutations-js-module")
+
+const mutations = createMutations({
+  mutations: myMutations,
+  subgraph: "my-subgraph",
+  node: "http://localhost:8080",
```

Comment:
This endpoint is used to query the metadata subgraph. I don't think we can get rid of this here because we need to know the query endpoint's root URL so we can append `/subgraphs` to it.

In the case of using `createQueryEngine`, the developer could just pass in the mutations & config there like so:
```typescript
let engine = createQueryEngine({
  subgraph: 'subgraph/a',
  mutations: {
    module: myMutations,
    config: { ... }
  }

  subgraphNameResolver: ...,
  indexersResolver: ...,
  indexNodeResolver: ...,
  queryPlanner: ...,
  queryExecutor: ...,
})
```

Thoughts?

Reply:
Yes, I think that works. Can we leave out the `subgraph` and `node` options from the design for now (see my earlier description about the data sources)?